File: xLiAd.DiagnosticLogCenter.Agent/TracingDiagnosticObserver.cs
using System;
using System.Collections.Generic;
using System.Text;
using xLiAd.DiagnosticLogCenter.Agent.DiagnosticProcessors;

namespace xLiAd.DiagnosticLogCenter.Agent
{
    internal class TracingDiagnosticObserver : IObserver<KeyValuePair<string, object>>
    {
        private readonly TracingDiagnosticMethodCollection _methodCollection;
        public TracingDiagnosticObserver(ITracingDiagnosticProcessor tracingDiagnosticProcessor)
        {
            _methodCollection = new TracingDiagnosticMethodCollection(tracingDiagnosticProcessor);
        }
        public void OnCompleted()
        {
        }

        public void OnError(Exception error)
        {
        }

        public void OnNext(KeyValuePair<string, object> value)
        {
            foreach (var method in _methodCollection)
            {
                try
                {
                    method.Invoke(value.Key, value.Value);
                }
                catch (Exception exception)
                {
                    
                }
            }
        }
    }
}

File: DiagnosticLogCenter-Java/src/main/java/com/xliad/diagnosticlogcenter/agent/TracingDiagnosticObserver.java
package com.xliad.diagnosticlogcenter.agent;

import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.ITracingDiagnosticProcessor;

public class TracingDiagnosticObserver {
    private final TracingDiagnosticMethodCollection methodCollection;

    public TracingDiagnosticObserver(ITracingDiagnosticProcessor processor) {
        this.methodCollection = new TracingDiagnosticMethodCollection(processor);
    }

    public void onNext(String eventName, Object value) {
        for (TracingDiagnosticMethod method : methodCollection) {
            try {
                method.invoke(eventName, value);
            } catch (Exception e) {
                // 忽略异常
            }
        }
    }
}